staking/slash/double-sign.go

```diff
@@ -1,52 +1,385 @@
+	oneHundredPercent               = numeric.NewDec(1)
+	fiftyPercent                    = numeric.MustNewDecFromStr("0.5")
+)
+
+// applySlashRate returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
+func applySlashRate(amount *big.Int, rate numeric.Dec) *big.Int {
+	return numeric.NewDecFromBigInt(
+		amount,
+	).Mul(rate).TruncateInt()
+}
+
+func delegatorSlashApply(
+	snapshot, current *staking.ValidatorWrapper,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	doubleSignEpoch *big.Int,
+	slashTrack *Application,
+) error {
+	for i, delegationSnapshot := range snapshot.Delegations {
+		slashDebt := applySlashRate(delegationSnapshot.Amount, rate)
+		halfOfSlashDebt := new(big.Int).Div(slashDebt, common.Big2)
+		snapshotAddr := delegationSnapshot.DelegatorAddress
+		for j, delegationNow := range current.Delegations {
+			if nowAmt := delegationNow.Amount; delegationNow.DelegatorAddress == snapshotAddr {
+				state.AddBalance(reporter, halfOfSlashDebt)
+				// NOTE only need to pay snitch here
+				slashTrack.TotalSnitchReward.Add(
+					slashTrack.TotalSnitchReward, halfOfSlashDebt,
+				)
+				// NOTE handle validator self delegation special case
+				if i == validatorsOwnDel && j == validatorsOwnDel {
+					// say my slash debt is 1.6, and my current amount is 1.2
+					// then while I can't drop below 1, I could still pay 0.2
+					// to bring my debt down to 1.4
+					if one := new(big.Int).SetUint64(denominations.One); nowAmt.Cmp(one) == 1 &&
+						slashDebt.Cmp(common.Big0) == 1 {
+						// have 1.1, but debt is 0.022
+						asMuchAsCouldPay := new(big.Int).Sub(nowAmt, one)
+						if asMuchAsCouldPay.Cmp(slashDebt) >= 0 {
+							mustStayAboveZeroAfterDebtReduce(
+								snapshot, current, slashDebt, slashDebt, slashTrack,
+							)
+							nowAmt.Sub(nowAmt, slashDebt)
+						}
+					}
+					// NOTE Assume did as much as could above, now check the undelegations
+					for _, undelegate := range delegationNow.Undelegations {
+						// the epoch matters, only those undelegation
+						// such that epoch>= doubleSignEpoch should be slashable
+						if undelegate.Epoch.Cmp(doubleSignEpoch) >= 0 {
+							if slashDebt.Cmp(common.Big0) <= 0 {
+								// paid off the slash debt
+								break
+							}
+
+							switch newBal := new(big.Int).Sub(
+								// My slash debt is 1.6 and my undelegate amount is 1.0
+								// so if (1.6 - 1.0) > 0
+								undelegate.Amount,
+								slashDebt); newBal.Cmp(common.Big0) {
+							case paidOffExact:
+								undelegate.Amount.SetInt64(0)
+							case haveEnoughToPayOff:
+								undelegate.Amount.Sub(undelegate.Amount, slashDebt)
+								mustStayAboveZeroAfterDebtReduce(
+									snapshot, current, slashDebt, slashDebt, slashTrack,
+								)
+								undelegate.Amount.Set(newBal)
+							case -1:
+								// But do have enough to pay off something at least
+								partialPayment := new(big.Int).Sub(slashDebt, new(big.Int).Abs(newBal))
```

Comment:
This partial payment is just the whole undelegation.Amount. No need for the calculation.